File: src/portfolio_manager/services/rebalance_execution_service.py
"""Rebalance execution service — converts recommendations to order intents."""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any

from portfolio_manager.models.rebalance import RebalanceRecommendation
from portfolio_manager.services.kis.kis_market_detector import is_domestic_ticker

_DEFAULT_OVERSEAS_EXCHANGE = "NASD"


@dataclass
class OrderIntent:
    """Standardized order request before sending to KIS API."""

    ticker: str
    side: str  # "buy" or "sell"
    quantity: int
    currency: str
    exchange: str | None = None  # overseas exchange code (NASD/NYSE/AMEX)


@dataclass
class OrderIntentResult:
    """Result of creating order intents: executable intents + skipped items."""

    intents: list[OrderIntent]
    skipped: list[OrderIntent]


@dataclass
class OrderExecutionResult:
    """Result of executing a single order."""

    intent: OrderIntent
    status: str  # "success", "failed", "skipped"
    message: str = ""
    raw_response: dict | None = None


@dataclass
class RebalanceExecutionResult:
    """Full result of executing rebalance orders."""

    intents: list[OrderIntent]
    skipped: list[OrderIntent]
    executions: list[OrderExecutionResult] | None = None
    sync_warning: str | None = None


class RebalanceExecutionService:
    """Converts rebalance recommendations into executable order intents."""

    def __init__(
        self,
        order_client: Any | None = None,
        execution_repository: Any | None = None,
        sync_service: Any | None = None,
    ):
        self._order_client = order_client
        self._execution_repo = execution_repository
        self._sync_service = sync_service

    def execute_rebalance_orders(
        self,
        recommendations: list[RebalanceRecommendation],
        *,
        dry_run: bool = True,
        exchange_map: dict[str, str | None] | None = None,
    ) -> RebalanceExecutionResult:
        intent_result = self.create_order_intents(
            recommendations, exchange_map=exchange_map
        )
        executions: list[OrderExecutionResult] = []
        if not dry_run and self._order_client:
            for intent in intent_result.intents:
                resp = self._order_client.place_order(
                    ticker=intent.ticker,
                    side=intent.side,
                    quantity=intent.quantity,
                    exchange=intent.exchange,
                )
                executions.append(
                    OrderExecutionResult(
                        intent=intent,
                        status="success" if resp.get("rt_cd") == "0" else "failed",
                        message=resp.get("msg1", ""),
                        raw_response=resp,
                    )
                )
        if not dry_run and self._execution_repo:
            for ex in executions:
                self._execution_repo.create(
                    ticker=ex.intent.ticker,
                    side=ex.intent.side,
                    quantity=ex.intent.quantity,
                    currency=ex.intent.currency,
                    exchange=ex.intent.exchange,
                    status=ex.status,
                    message=ex.message,
                    raw_response=ex.raw_response,
                )
            for sk in intent_result.skipped:
                self._execution_repo.create(
                    ticker=sk.ticker,
                    side=sk.side,
                    quantity=sk.quantity,
                    currency=sk.currency,
                    exchange=sk.exchange,
                    status="skipped",
                    message="",
                    raw_response=None,
                )
        sync_warning: str | None = None
        if not dry_run and self._sync_service:
            try:
                self._sync_service.sync_account()
            except Exception as e:
                sync_warning = str(e)
        return RebalanceExecutionResult(
            intents=intent_result.intents,
            skipped=intent_result.skipped,
            executions=executions,
            sync_warning=sync_warning,
        )

    def create_order_intents(
        self,
        recommendations: list[RebalanceRecommendation],
        *,
        exchange_map: dict[str, str | None] | None = None,
    ) -> OrderIntentResult:
        intents: list[OrderIntent] = []
        skipped: list[OrderIntent] = []
        ex_map = exchange_map or {}
        for rec in recommendations:
            qty = int(math.floor(rec.quantity)) if rec.quantity is not None else 0
            exchange: str | None = None
            if not is_domestic_ticker(rec.ticker):
                exchange = ex_map.get(rec.ticker) or _DEFAULT_OVERSEAS_EXCHANGE
            intent = OrderIntent(
                ticker=rec.ticker,
                side=rec.action.value,
                quantity=qty,
                currency=rec.currency
                if rec.currency
                else "KRW"
                if is_domestic_ticker(rec.ticker)
                else "USD",
                exchange=exchange,
            )
            if qty == 0:
                skipped.append(intent)
            else:
                intents.append(intent)
        intents.sort(key=lambda i: i.side != "sell")
        return OrderIntentResult(intents=intents, skipped=skipped)

File: internal/services/rebalance_execution_service.go
package services

import (
	"context"
	"math"

	"github.com/kadragon/portfolio-manager/internal/models"
)

const defaultOverseasExchange = "NASD"

// OrderClient places a single order through the configured brokerage route.
type OrderClient interface {
	PlaceOrder(intent models.OrderIntent) (map[string]any, error)
}

// ExecutionRepo persists order execution records.
type ExecutionRepo interface {
	Create(
		ctx context.Context,
		ticker, side string,
		quantity int,
		currency, status, message string,
		exchange string,
		rawResponse map[string]any,
	) (models.OrderExecutionRecord, error)
}

// SyncService syncs account state after order execution.
type SyncService interface {
	SyncAccount() error
}

// RebalanceExecutionService converts recommendations into executable order intents.
type RebalanceExecutionService struct {
	orderClient OrderClient
	execRepo    ExecutionRepo
	syncService SyncService
}

// NewRebalanceExecutionService builds the service. All deps are optional.
func NewRebalanceExecutionService(
	orderClient OrderClient,
	execRepo ExecutionRepo,
	syncService SyncService,
) *RebalanceExecutionService {
	return &RebalanceExecutionService{
		orderClient: orderClient,
		execRepo:    execRepo,
		syncService: syncService,
	}
}

// OrderIntentResult holds executable intents and skipped (qty=0) items.
type OrderIntentResult struct {
	Intents []models.OrderIntent
	Skipped []models.OrderIntent
}

// CreateOrderIntents converts recommendations to executable intents.
// exchangeMap is an optional ticker→exchange override (nil = use defaults).
// Sells are ordered before buys; qty is floored; qty=0 → skipped.
func (s *RebalanceExecutionService) CreateOrderIntents(
	recs []models.RebalanceRecommendation,
	exchangeMap map[string]string,
) OrderIntentResult {
	var intents []models.OrderIntent
	var skipped []models.OrderIntent

	for _, rec := range recs {
		var qty int
		if rec.Quantity != nil {
			qty = int(math.Floor(rec.Quantity.InexactFloat64()))
		}

		exchange := ""
		if !isDomesticTicker(rec.Ticker) {
			if exchangeMap != nil {
				if ex, ok := exchangeMap[rec.Ticker]; ok && ex != "" {
					exchange = ex
				} else {
					exchange = defaultOverseasExchange
				}
			} else {
				exchange = defaultOverseasExchange
			}
		}

		currency := rec.Currency
		if currency == "" {
			if isDomesticTicker(rec.Ticker) {
				currency = "KRW"
			} else {
				currency = "USD"
			}
		}

		intent := models.OrderIntent{
			Ticker:      rec.Ticker,
			Side:        string(rec.Action),
			Quantity:    qty,
			Currency:    currency,
			Exchange:    exchange,
			StockName:   rec.StockName,
			AccountID:   rec.AccountID,
			AccountName: rec.AccountName,
			Amount:      rec.Amount,
		}
		if qty == 0 {
			skipped = append(skipped, intent)
		} else {
			intents = append(intents, intent)
		}
	}

	// sells before buys (stable sort preserves original order within each group)
	stable := make([]models.OrderIntent, 0, len(intents))
	for _, in := range intents {
		if in.Side == "sell" {
			stable = append(stable, in)
		}
	}
	for _, in := range intents {
		if in.Side == "buy" {
			stable = append(stable, in)
		}
	}
	return OrderIntentResult{Intents: stable, Skipped: skipped}
}

// ExecuteRebalanceOrders executes orders and persists results.
// If dryRun=true, no API calls are made and only intents are returned.
func (s *RebalanceExecutionService) ExecuteRebalanceOrders(
	recs []models.RebalanceRecommendation,
	dryRun bool,
	exchangeMap map[string]string,
) models.RebalanceExecutionResult {
	intentResult := s.CreateOrderIntents(recs, exchangeMap)
	executable, deferred := splitExecutableIntents(intentResult.Intents)

	var executions []models.OrderExecutionResult
	if !dryRun && s.orderClient != nil {
		for _, intent := range executable {
			resp, err := s.orderClient.PlaceOrder(intent)
			status := "success"
			msg := ""
			if err != nil {
				status = "failed"
				msg = err.Error()
			} else if resp != nil {
				if rtCd, ok := resp["rt_cd"].(string); ok && rtCd != "0" {
					status = "failed"
				}
				if m, ok := resp["msg1"].(string); ok {
					msg = m
				}
			}
			executions = append(executions, models.OrderExecutionResult{
				Intent:      intent,
				Status:      status,
				Message:     msg,
				RawResponse: resp,
			})
		}
	}

	if !dryRun && s.execRepo != nil {
		for _, ex := range executions {
			_, _ = s.execRepo.Create(
				context.Background(),
				ex.Intent.Ticker, ex.Intent.Side,
				ex.Intent.Quantity,
				ex.Intent.Currency, ex.Status, ex.Message,
				ex.Intent.Exchange,
				ex.RawResponse,
			)
		}
		for _, sk := range intentResult.Skipped {
			_, _ = s.execRepo.Create(
				context.Background(),
				sk.Ticker, sk.Side,
				sk.Quantity,
				sk.Currency, "skipped", "",
				sk.Exchange,
				nil,
			)
		}
		for _, d := range deferred {
			_, _ = s.execRepo.Create(
				context.Background(),
				d.Ticker, d.Side,
				d.Quantity,
				d.Currency, "deferred", "deferred until sell orders are confirmed and account state is synced",
				d.Exchange,
				nil,
			)
		}
	}

	syncWarning := ""
	if !dryRun && s.syncService != nil {
		if err := s.syncService.SyncAccount(); err != nil {
			syncWarning = err.Error()
		}
	}

	return models.RebalanceExecutionResult{
		Intents:     intentResult.Intents,
		Skipped:     intentResult.Skipped,
		Deferred:    deferred,
		Executions:  executions,
		SyncWarning: syncWarning,
	}
}

func splitExecutableIntents(intents []models.OrderIntent) (executable, deferred []models.OrderIntent) {
	sellAccounts := make(map[string]bool)
	for _, intent := range intents {
		if intent.Side == "sell" {
			sellAccounts[intent.AccountID.String()] = true
		}
	}
	if len(sellAccounts) == 0 {
		return intents, nil
	}
	for _, intent := range intents {
		if intent.Side == "buy" && sellAccounts[intent.AccountID.String()] {
			deferred = append(deferred, intent)
			continue
		}
		executable = append(executable, intent)
	}
	return executable, deferred
}
